### PR TITLE
ARROW-7636: [Python] Clean-up the pyarrow.dataset.partitioning() API

### PR DESCRIPTION
**Pull request by jorisvandenbossche**

### CODE REVIEW

File: python/pyarrow/dataset.py [hunk 1]
```diff
@@ -81,9 +81,13 @@ def partitioning(field_names=None, flavor=None):
...
-    field_names : pyarrow.Schema or list of str
-        The schema that describes the partitions present in the file path. If
-        a list of strings (field names) is passed, the schema's types are
+    schema : pyarrow.Schema, default None
+        The schema that describes the partitions present in the file path.
+        If not specified, and `field_names` and/or `flavor` are specified,
+        the schema will be inferred from the file path (and a
+        PartitioningFactory is returned).
+    field_names :  list of str, default None
+        A list of strings (field names). If specified, the schema's types are
```
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py]: @bkietz @kszucs following up on https://github.com/apache/arrow/pull/6022#discussion_r367016454,  I went with splitting the keyword into two separate ones. 
But I can also rename it to `schema_or_field_names` and keep it as a single one (it's long, but you can use it positionally), if that has stronger preference. 

I think having separate ones is a bit cleaner, and easier to explain.

Comment (kszucs) [reply python/pyarrow/dataset.py]: And it is easier to merge than separate later.

File: python/pyarrow/dataset.py [hunk 2]
```diff
@@ -121,27 +125,34 @@ def partitioning(field_names=None, flavor=None):
...
-            return HivePartitioning(field_names)
-        elif field_names is None:
-            return HivePartitioning.discover()
+        if field_names is not None:
+            raise ValueError("Cannot specify 'field_names' for flavor 'hive'")
+        elif schema is not None:
+            if isinstance(schema, pa.Schema):
+                return HivePartitioning(schema)
+            else:
+                raise ValueError(
```
Comment (kszucs) [inline python/pyarrow/dataset.py]: This is why I found useful the ability to restrict the HivePartitioning for certain field names. Perhaps I'm only interested in a subset of the partitioning fields, and it'd be easier to define them here.

@bkietz created an issue where we can discuss https://issues.apache.org/jira/browse/ARROW-7646

**Review by kszucs — APPROVED** [review]
LGTM

**Pull request closed without merging**